src/module/archmage.js

```diff
@@ -1143,9 +1123,56 @@ Hooks.on('dropCanvasData', async (canvas, data) => {
         }
       }
     }
-    return token.actor.createEmbeddedDocuments("ActiveEffect", [effectData]);
+    return await _applyAEDurationDialog(actor, effectData, data.ends, data.source, token);
   }
-});
+}
+
+async function _applyAEDurationDialog(actor, effectData, duration, source, token=undefined) {
+  if (event.shiftKey) {
+    if ( token ) return token.toggleEffect(effectData, {active: true});
+    else return actor.createEmbeddedDocuments("ActiveEffect", [effectData]);
+  }
+  // Render modal dialog
+  const sourceActor = await fromUuid(source);
+  const template = 'systems/archmage/templates/chat/apply-AE.html';
+  let dialogData = {
+    effectName: effectData.name,
+    sourceName: sourceActor?.name ?? "",
+    defaultDuration: duration,
+    durations: CONFIG.ARCHMAGE.effectDurationTypes
+  };
+  let do_apply = false;
+
+  renderTemplate(template, dialogData).then(dlg => {
+    new Dialog({
+      title: game.i18n.localize("ARCHMAGE.CHAT.applyAETitle"),
+      content: dlg,
+      buttons: {
+        apply: {
+          label: game.i18n.localize("ARCHMAGE.CHAT.Apply"),
+          callback: () => { do_apply = true; }
+        },
+        pen1: {
+          label: game.i18n.localize("ARCHMAGE.CHAT.Cancel"),
+          callback: () => { do_apply = false; }
+        },
+      },
+      default: 'apply',
+      close: html => {
+        if (do_apply) {
+          duration = html.find('[name="duration"]').val();
+          let options = []
+          if (['StartOfNextSourceTurn', 'EndOfNextSourceTurn'].includes(duration)) {
+            options = {sourceTurnUuid: source};
+          }
+          game.archmage.MacroUtils.setDuration(effectData, duration, options);
+          if ( token ) return token._object.toggleEffect(effectData, {active: true});
+          else return actor.createEmbeddedDocuments("ActiveEffect", [effectData]);
+        }
+      }
```

Comment:
Any reason you don't just put the contents of this `if` into the `apply` callback?

Reply:
With the current version of the code, none really. It began with a different idea and I didn't quite cleaned it up yet.